---
title: 'Excepciones y control de excepciones: Guía de programación de C#'
description: Obtenga más información sobre las excepciones y cómo controlarlas. Estas características de C# ayudan a lidiar con aquellas situaciones excepcionales o inesperadas que se dan cuando un programa se está ejecutando.
ms.date: 07/20/2015
helpviewer_keywords:
- exception handling [C#]
- exceptions [C#]
- C# language, exceptions
ms.assetid: 0001887f-4fa2-47e2-8034-2819477e2344
ms.openlocfilehash: 4be327be4e183d60e440358254aeb9c68fcf25ca
ms.sourcegitcommit: 6f58a5f75ceeb936f8ee5b786e9adb81a9a3bee9
ms.translationtype: HT
ms.contentlocale: es-ES
ms.lasthandoff: 07/28/2020
ms.locfileid: "87303392"
---
# <a name="exceptions-and-exception-handling-c-programming-guide"></a><span data-ttu-id="52c51-104">Excepciones y control de excepciones (Guía de programación de C#)</span><span class="sxs-lookup"><span data-stu-id="52c51-104">Exceptions and Exception Handling (C# Programming Guide)</span></span>

<span data-ttu-id="52c51-105">Las características de control de excepciones del lenguaje C# le ayudan a afrontar cualquier situación inesperada o excepcional que se produce cuando se ejecuta un programa.</span><span class="sxs-lookup"><span data-stu-id="52c51-105">The C# language's exception handling features help you deal with any unexpected or exceptional situations that occur when a program is running.</span></span> <span data-ttu-id="52c51-106">El control de excepciones usa las palabras clave `try`, `catch` y `finally` para intentar realizar acciones que pueden no completarse correctamente, para controlar errores cuando decide que es razonable hacerlo y para limpiar recursos más adelante.</span><span class="sxs-lookup"><span data-stu-id="52c51-106">Exception handling uses the `try`, `catch`, and `finally` keywords to try actions that may not succeed, to handle failures when you decide that it is reasonable to do so, and to clean up resources afterward.</span></span> <span data-ttu-id="52c51-107">Las excepciones las puede generar Common Language Runtime (CLR), .NET, bibliotecas de terceros o el código de aplicación.</span><span class="sxs-lookup"><span data-stu-id="52c51-107">Exceptions can be generated by the common language runtime (CLR), by .NET or third-party libraries, or by application code.</span></span> <span data-ttu-id="52c51-108">Las excepciones se crean mediante el uso de la palabra clave `throw`.</span><span class="sxs-lookup"><span data-stu-id="52c51-108">Exceptions are created by using the `throw` keyword.</span></span>

<span data-ttu-id="52c51-109">En muchos casos, una excepción la puede no producir un método al que el código ha llamado directamente, sino otro método más bajo en la pila de llamadas.</span><span class="sxs-lookup"><span data-stu-id="52c51-109">In many cases, an exception may be thrown not by a method that your code has called directly, but by another method further down in the call stack.</span></span> <span data-ttu-id="52c51-110">Cuando esto sucede, CLR desenredar la pila, busca un método con un bloque `catch` para el tipo de excepción específico y ejecuta el primer bloque `catch` que encuentra.</span><span class="sxs-lookup"><span data-stu-id="52c51-110">When this happens, the CLR will unwind the stack, looking for a method with a `catch` block for the specific exception type, and it will execute the first such `catch` block that if finds.</span></span> <span data-ttu-id="52c51-111">Si no encuentra ningún bloque `catch` adecuado en cualquier parte de la pila de llamadas, finalizará el proceso y mostrará un mensaje al usuario.</span><span class="sxs-lookup"><span data-stu-id="52c51-111">If it finds no appropriate `catch` block anywhere in the call stack, it will terminate the process and display a message to the user.</span></span>

<span data-ttu-id="52c51-112">En este ejemplo, un método prueba a hacer la división entre cero y detecta el error.</span><span class="sxs-lookup"><span data-stu-id="52c51-112">In this example, a method tests for division by zero and catches the error.</span></span> <span data-ttu-id="52c51-113">Sin el control de excepciones, este programa finalizaría con un error **DivideByZeroException no controlada**.</span><span class="sxs-lookup"><span data-stu-id="52c51-113">Without the exception handling, this program would terminate with a **DivideByZeroException was unhandled** error.</span></span>

[!code-csharp[csProgGuideExceptions#18](~/samples/snippets/csharp/VS_Snippets_VBCSharp/csProgGuideExceptions/CS/Exceptions.cs#18)]

## <a name="exceptions-overview"></a><span data-ttu-id="52c51-114">Información general sobre excepciones</span><span class="sxs-lookup"><span data-stu-id="52c51-114">Exceptions Overview</span></span>

<span data-ttu-id="52c51-115">Las excepciones tienen las siguientes propiedades:</span><span class="sxs-lookup"><span data-stu-id="52c51-115">Exceptions have the following properties:</span></span>

- <span data-ttu-id="52c51-116">Las excepciones son tipos que derivan en última instancia de `System.Exception`.</span><span class="sxs-lookup"><span data-stu-id="52c51-116">Exceptions are types that all ultimately derive from `System.Exception`.</span></span>
- <span data-ttu-id="52c51-117">Use un bloque `try` alrededor de las instrucciones que pueden producir excepciones.</span><span class="sxs-lookup"><span data-stu-id="52c51-117">Use a `try` block around the statements that might throw exceptions.</span></span>
- <span data-ttu-id="52c51-118">Una vez que se produce una excepción en el bloque `try`, el flujo de control salta al primer controlador de excepciones asociado que está presente en cualquier parte de la pila de llamadas.</span><span class="sxs-lookup"><span data-stu-id="52c51-118">Once an exception occurs in the `try` block, the flow of control jumps to the first associated exception handler that is present anywhere in the call stack.</span></span> <span data-ttu-id="52c51-119">En C#, la palabra clave `catch` se utiliza para definir un controlador de excepciones.</span><span class="sxs-lookup"><span data-stu-id="52c51-119">In C#, the `catch` keyword is used to define an exception handler.</span></span>
- <span data-ttu-id="52c51-120">Si no hay ningún controlador de excepciones para una excepción determinada, el programa deja de ejecutarse con un mensaje de error.</span><span class="sxs-lookup"><span data-stu-id="52c51-120">If no exception handler for a given exception is present, the program stops executing with an error message.</span></span>
- <span data-ttu-id="52c51-121">No detecte una excepción a menos que pueda controlarla y dejar la aplicación en un estado conocido.</span><span class="sxs-lookup"><span data-stu-id="52c51-121">Do not catch an exception unless you can handle it and leave the application in a known state.</span></span> <span data-ttu-id="52c51-122">Si se detecta `System.Exception`, reinícielo con la palabra clave `throw` al final del bloque `catch`.</span><span class="sxs-lookup"><span data-stu-id="52c51-122">If you catch `System.Exception`, re-throw it using the `throw` keyword at the end of the `catch` block.</span></span>
- <span data-ttu-id="52c51-123">Si un bloque `catch` define una variable de excepción, puede utilizarla para obtener más información sobre el tipo de excepción que se ha producido.</span><span class="sxs-lookup"><span data-stu-id="52c51-123">If a `catch` block defines an exception variable, you can use it to obtain more information about the type of exception that occurred.</span></span>
- <span data-ttu-id="52c51-124">Las excepciones puede generarlas explícitamente un programa con la palabra clave `throw`.</span><span class="sxs-lookup"><span data-stu-id="52c51-124">Exceptions can be explicitly generated by a program by using the `throw` keyword.</span></span>
- <span data-ttu-id="52c51-125">Los objetos de excepción contienen información detallada sobre el error, como el estado de la pila de llamadas y una descripción de texto del error.</span><span class="sxs-lookup"><span data-stu-id="52c51-125">Exception objects contain detailed information about the error, such as the state of the call stack and a text description of the error.</span></span>
- <span data-ttu-id="52c51-126">El código de un bloque `finally` se ejecuta incluso si se produce una excepción.</span><span class="sxs-lookup"><span data-stu-id="52c51-126">Code in a `finally` block is executed even if an exception is thrown.</span></span> <span data-ttu-id="52c51-127">Use un bloque `finally` para liberar recursos, por ejemplo, para cerrar las secuencias o los archivos que se abrieron en el bloque `try`.</span><span class="sxs-lookup"><span data-stu-id="52c51-127">Use a `finally` block to release resources, for example to close any streams or files that were opened in the `try` block.</span></span>
- <span data-ttu-id="52c51-128">Las excepciones administradas de .NET se implementan en el mecanismo de control de excepciones estructurado de Win32.</span><span class="sxs-lookup"><span data-stu-id="52c51-128">Managed exceptions in .NET are implemented on top of the Win32 structured exception handling mechanism.</span></span> <span data-ttu-id="52c51-129">Para más información, vea [Control de excepciones estructurado (C/C++)](/cpp/cpp/structured-exception-handling-c-cpp) y [A Crash Course on the Depths of Win32 Structured Exception Handling](http://bytepointer.com/resources/pietrek_crash_course_depths_of_win32_seh.htm) (Curso intensivo sobre los aspectos específicos del control de excepciones estructurado de Win32).</span><span class="sxs-lookup"><span data-stu-id="52c51-129">For more information, see [Structured Exception Handling (C/C++)](/cpp/cpp/structured-exception-handling-c-cpp) and [A Crash Course on the Depths of Win32 Structured Exception Handling](http://bytepointer.com/resources/pietrek_crash_course_depths_of_win32_seh.htm).</span></span>

## <a name="related-sections"></a><span data-ttu-id="52c51-130">Secciones relacionadas</span><span class="sxs-lookup"><span data-stu-id="52c51-130">Related Sections</span></span>

<span data-ttu-id="52c51-131">Para obtener más información sobre las excepciones y el control de excepciones, vea los artículos siguientes:</span><span class="sxs-lookup"><span data-stu-id="52c51-131">See the following articles for more information about exceptions and exception handling:</span></span>

- [<span data-ttu-id="52c51-132">Utilizar excepciones</span><span class="sxs-lookup"><span data-stu-id="52c51-132">Using Exceptions</span></span>](using-exceptions.md)
- [<span data-ttu-id="52c51-133">Control de excepciones</span><span class="sxs-lookup"><span data-stu-id="52c51-133">Exception Handling</span></span>](exception-handling.md)
- [<span data-ttu-id="52c51-134">Crear y producir excepciones</span><span class="sxs-lookup"><span data-stu-id="52c51-134">Creating and Throwing Exceptions</span></span>](creating-and-throwing-exceptions.md)
- [<span data-ttu-id="52c51-135">Excepciones generadas por el compilador</span><span class="sxs-lookup"><span data-stu-id="52c51-135">Compiler-Generated Exceptions</span></span>](compiler-generated-exceptions.md)
- [<span data-ttu-id="52c51-136">Procedimiento para controlar una excepción mediante Try y Catch (Guía de programación de C#)</span><span class="sxs-lookup"><span data-stu-id="52c51-136">How to handle an exception using try/catch (C# Programming Guide)</span></span>](how-to-handle-an-exception-using-try-catch.md)
- [<span data-ttu-id="52c51-137">Procedimiento para ejecutar código de limpieza mediante finally</span><span class="sxs-lookup"><span data-stu-id="52c51-137">How to execute cleanup code using finally</span></span>](how-to-execute-cleanup-code-using-finally.md)
- [<span data-ttu-id="52c51-138">Procedimiento para detectar excepciones no compatibles con CLS</span><span class="sxs-lookup"><span data-stu-id="52c51-138">How to catch a non-CLS exception</span></span>](how-to-catch-a-non-cls-exception.md)

## <a name="c-language-specification"></a><span data-ttu-id="52c51-139">Especificación del lenguaje C#</span><span class="sxs-lookup"><span data-stu-id="52c51-139">C# Language Specification</span></span>

<span data-ttu-id="52c51-140">Para obtener más información, consulte la sección [Excepciones](~/_csharplang/spec/exceptions.md) de [Especificación del lenguaje C#](/dotnet/csharp/language-reference/language-specification/introduction).</span><span class="sxs-lookup"><span data-stu-id="52c51-140">For more information, see [Exceptions](~/_csharplang/spec/exceptions.md) in the [C# Language Specification](/dotnet/csharp/language-reference/language-specification/introduction).</span></span> <span data-ttu-id="52c51-141">La especificación del lenguaje es la fuente definitiva de la sintaxis y el uso de C#.</span><span class="sxs-lookup"><span data-stu-id="52c51-141">The language specification is the definitive source for C# syntax and usage.</span></span>

## <a name="see-also"></a><span data-ttu-id="52c51-142">Vea también</span><span class="sxs-lookup"><span data-stu-id="52c51-142">See also</span></span>

- <xref:System.SystemException>
- [<span data-ttu-id="52c51-143">Guía de programación de C#</span><span class="sxs-lookup"><span data-stu-id="52c51-143">C# Programming Guide</span></span>](../index.md)
- [<span data-ttu-id="52c51-144">Palabras clave de C#</span><span class="sxs-lookup"><span data-stu-id="52c51-144">C# Keywords</span></span>](../../language-reference/keywords/index.md)
- [<span data-ttu-id="52c51-145">throw</span><span class="sxs-lookup"><span data-stu-id="52c51-145">throw</span></span>](../../language-reference/keywords/throw.md)
- [<span data-ttu-id="52c51-146">try-catch</span><span class="sxs-lookup"><span data-stu-id="52c51-146">try-catch</span></span>](../../language-reference/keywords/try-catch.md)
- [<span data-ttu-id="52c51-147">try-finally</span><span class="sxs-lookup"><span data-stu-id="52c51-147">try-finally</span></span>](../../language-reference/keywords/try-finally.md)
- [<span data-ttu-id="52c51-148">try-catch-finally</span><span class="sxs-lookup"><span data-stu-id="52c51-148">try-catch-finally</span></span>](../../language-reference/keywords/try-catch-finally.md)
- [<span data-ttu-id="52c51-149">Excepciones</span><span class="sxs-lookup"><span data-stu-id="52c51-149">Exceptions</span></span>](../../../standard/exceptions/index.md)
